File: Source/LefticeRuntime/Private/CoreUObject/EnumProperty.cpp
// Copyright (c) NextTurn.
// See LICENSE.txt in the project root for more information.

#include "LefticeRuntime/Private/Exports.hpp"

#include "LefticeRuntime/Private/Accessor.hpp"
#include "LefticeRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/EnumProperty.h"

ENABLE_WARNINGS;

DEFINE_OFFSET_ACCESSOR(UEnumProperty, Enum);
DEFINE_OFFSET_ACCESSOR(UEnumProperty, UnderlyingProp);

#define TYPE_EXPORT_MANAGED_TYPE_NAME EnumProperty

TYPE_EXPORT_START

const FMemberSymbol Members[] =
{
	EXPORT_OFFSET(UEnumProperty, MetaEnum, Enum),
	EXPORT_OFFSET(UEnumProperty, UnderlyingProperty, UnderlyingProp),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;

File: Managed/Leftice.Runtime/CoreUObject/EnumProperty.cs
// Copyright (c) NextTurn.
// See LICENSE.txt in the project root for more information.

using System;
using Leftice.Processors;

namespace Unreal
{
    public sealed class EnumProperty : Property
    {
        internal EnumProperty(IntPtr pointer) : base(pointer) { }

        public Enum MetaEnum => Create<Enum>(NativeMethods.GetMetaEnum(this.pointer));

        public NumericProperty UnderlyingProperty => Create<NumericProperty>(NativeMethods.GetUnderlyingProperty(this.pointer));

        private static new class NativeMethods
        {
            [ReadOffset]
            public static extern IntPtr GetMetaEnum(IntPtr property);

            [ReadOffset]
            public static extern IntPtr GetUnderlyingProperty(IntPtr property);
        }
    }
}
